File: scripts/models/linearRegression.js
// scripts/models/linearRegression.js
// Linear Regression with Batch Gradient Descent & L2 Regularization (Ridge)
// Implemented strictly from scratch in vanilla JavaScript.

import { matVecMul, matTransposeVecMul, rSquared, mse, rmse } from '../utils/math.js';

export class LinearRegression {
  /**
   * @param {Object} options
   * @param {number} [options.learningRate=0.05]
   * @param {number} [options.epochs=200]
   * @param {boolean} [options.useRegularization=false]
   * @param {number} [options.lambda=1.0] - L2 regularization penalty parameter
   */
  constructor(options = {}) {
    this.learningRate = options.learningRate ?? 0.05;
    this.epochs = options.epochs ?? 200;
    this.useRegularization = options.useRegularization ?? false;
    this.lambda = options.lambda ?? 1.0;

    this.weights = []; // Length d
    this.bias = 0;
    this.history = []; // Array of { epoch, cost }
    this.isDiverged = false;
    this.currentEpoch = 0;
    this.isTraining = false;
  }

  /**
   * Initializes weights and bias to zeros or small random numbers.
   * @param {number} numFeatures 
   */
  initializeWeights(numFeatures) {
    this.weights = new Array(numFeatures).fill(0);
    this.bias = 0;
    this.history = [];
    this.isDiverged = false;
    this.currentEpoch = 0;
  }

  /**
   * Forward pass: computes predictions y_hat = X * w + b.
   * @param {number[][]} X - Matrix of shape [m, d]
   * @returns {number[]} Vector of predictions of length m
   */
  predictRaw(X) {
    const xw = matVecMul(X, this.weights);
    const m = X.length;
    const yHat = new Array(m);
    for (let i = 0; i < m; i++) {
      yHat[i] = xw[i] + this.bias;
    }
    return yHat;
  }

  /**
   * Computes Mean Squared Error with optional L2 regularization.
   * J(w, b) = (1 / 2m) * sum((y_hat - y)^2) + (lambda / 2m) * sum(w^2)
   * @param {number[]} yHat 
   * @param {number[]} yTrue 
   * @returns {number}
   */
  computeCost(yHat, yTrue) {
    const m = yTrue.length;
    if (m === 0) return 0;

    let sumSqErr = 0;
    for (let i = 0; i < m; i++) {
      const err = yHat[i] - yTrue[i];
      sumSqErr += err * err;
    }

    let cost = sumSqErr / (2 * m);

    if (this.useRegularization && this.lambda > 0) {
      let l2Sum = 0;
      for (let j = 0; j < this.weights.length; j++) {
        l2Sum += this.weights[j] * this.weights[j];
      }
      cost += (this.lambda / (2 * m)) * l2Sum;
    }

    return cost;
  }

  /**
   * Executes a single epoch of vectorized batch gradient descent.
   * @param {number[][]} X - Shape [m, d]
   * @param {number[]} y - Shape [m]
   * @returns {{ epoch: number, cost: number }}
   */
  trainStep(X, y) {
    const m = X.length;
    const d = X[0].length;

    if (this.weights.length !== d) {
      this.initializeWeights(d);
    }

    // 1. Forward hypothesis: y_hat = X * w + b
    const yHat = this.predictRaw(X);

    // 2. Compute cost
    const cost = this.computeCost(yHat, y);

    // Guard against NaN or explosion
    if (isNaN(cost) || !isFinite(cost) || cost > 1e10) {
      this.isDiverged = true;
      return { epoch: this.currentEpoch, cost: NaN, diverged: true };
    }

    // 3. Compute error vector: (y_hat - y)
    const errors = new Array(m);
    let sumError = 0;
    for (let i = 0; i < m; i++) {
      const err = yHat[i] - y[i];
      errors[i] = err;
      sumError += err;
    }

    // 4. Compute gradient w.r.t weights: (1 / m) * X^T * errors + (lambda / m) * w
    const gradW = matTransposeVecMul(X, errors);
    const regFactor = (this.useRegularization && this.lambda > 0) ? (this.lambda / m) : 0;

    for (let j = 0; j < d; j++) {
      const g = (gradW[j] / m) + (regFactor * this.weights[j]);
      this.weights[j] -= this.learningRate * g;
    }

    // 5. Compute gradient w.r.t bias: (1 / m) * sum(errors)
    const gradB = sumError / m;
    this.bias -= this.learningRate * gradB;

    this.currentEpoch++;
    const stepRecord = { epoch: this.currentEpoch, cost, diverged: false };
    this.history.push(stepRecord);

    return stepRecord;
  }

  /**
   * Synchronously trains all epochs (ideal for quick tests/benchmarks).
   * @param {number[][]} X 
   * @param {number[]} y 
   */
  train(X, y) {
    const d = X[0].length;
    this.initializeWeights(d);
    for (let e = 0; e < this.epochs; e++) {
      const res = this.trainStep(X, y);
      if (res.diverged) break;
    }
  }

  /**
   * Predict single standardized feature vector.
   * @param {number[]} xStd 
   * @returns {number} Normalized prediction
   */
  predictSample(xStd) {
    let sum = this.bias;
    for (let j = 0; j < this.weights.length; j++) {
      sum += this.weights[j] * xStd[j];
    }
    return sum;
  }

  /**
   * Full evaluation on test dataset.
   * Computes R^2, MSE, RMSE (both normalized and unscaled in original USD).
   * @param {number[][]} XTestStd 
   * @param {number[]} yTestNorm 
   * @param {number[]} yTestRaw 
   * @param {{ mean: number, stdDev: number }} targetStats 
   */
  evaluate(XTestStd, yTestNorm, yTestRaw, targetStats) {
    const yPredNorm = this.predictRaw(XTestStd);
    const r2Norm = rSquared(yTestNorm, yPredNorm);
    const mseNorm = mse(yTestNorm, yPredNorm);

    // Unscale to raw USD
    const yPredRaw = yPredNorm.map(val => val * targetStats.stdDev + targetStats.mean);
    const r2Raw = rSquared(yTestRaw, yPredRaw);
    const mseRaw = mse(yTestRaw, yPredRaw);
    const rmseRaw = rmse(yTestRaw, yPredRaw);

    return {
      r2: r2Raw,
      mse: mseRaw,
      rmse: rmseRaw,
      normalizedMSE: mseNorm,
      weights: [...this.weights],
      bias: this.bias
    };
  }
}

File: ml/linear_regression.py
"""
ml/linear_regression.py
From-scratch Linear Regression using vectorized Batch Gradient Descent with NumPy.
Supports L2 Regularization (Ridge) and divergence guards.
"""

from typing import Dict, Any, List, Optional
import numpy as np


class LinearRegression:
    def __init__(
        self,
        learning_rate: float = 0.05,
        epochs: int = 200,
        use_regularization: bool = False,
        lambda_reg: float = 0.0
    ):
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.use_regularization = use_regularization
        self.lambda_reg = lambda_reg

        self.weights: Optional[np.ndarray] = None
        self.bias: float = 0.0
        self.history: List[Dict[str, Any]] = []
        self.is_diverged: bool = False

    def initialize_weights(self, num_features: int) -> None:
        self.weights = np.zeros(num_features, dtype=np.float64)
        self.bias = 0.0
        self.history = []
        self.is_diverged = False

    def predict_norm(self, X: np.ndarray) -> np.ndarray:
        """Predicts normalized target: y_hat = X @ w + b."""
        if self.weights is None:
            raise ValueError("Model has not been trained yet.")
        return X @ self.weights + self.bias

    def compute_cost(self, y_hat: np.ndarray, y_true: np.ndarray) -> float:
        """
        Computes Mean Squared Error with optional L2 regularization on weights.
        J(w, b) = (1 / 2m) * sum((y_hat - y)^2) + (lambda / 2m) * sum(w^2)
        """
        m = len(y_true)
        if m == 0:
            return 0.0

        errors = y_hat - y_true
        cost = np.sum(errors ** 2) / (2.0 * m)

        if self.use_regularization and self.lambda_reg > 0:
            l2_penalty = (self.lambda_reg / (2.0 * m)) * np.sum(self.weights ** 2)
            cost += l2_penalty

        return float(cost)

    def train_step(self, X: np.ndarray, y: np.ndarray, epoch_num: int) -> Dict[str, Any]:
        """Performs a single vectorized batch gradient descent update step."""
        m, d = X.shape
        if self.weights is None or len(self.weights) != d:
            self.initialize_weights(d)

        y_hat = self.predict_norm(X)
        cost = self.compute_cost(y_hat, y)

        if np.isnan(cost) or np.isinf(cost) or cost > 1e10:
            self.is_diverged = True
            return {"epoch": epoch_num, "cost": float("nan"), "diverged": True}

        errors = y_hat - y

        # Vectorized gradient calculation: (1 / m) * (X^T @ errors) + (lambda / m) * w
        grad_w = (X.T @ errors) / m
        if self.use_regularization and self.lambda_reg > 0:
            grad_w += (self.lambda_reg / m) * self.weights

        grad_b = float(np.sum(errors) / m)

        # Gradient descent update
        self.weights -= self.learning_rate * grad_w
        self.bias -= self.learning_rate * grad_b

        step_record = {"epoch": epoch_num, "cost": float(cost), "diverged": False}
        self.history.append(step_record)
        return step_record

    def fit(self, X: np.ndarray, y: np.ndarray) -> "LinearRegression":
        """Trains the model across all configured epochs."""
        m, d = X.shape
        self.initialize_weights(d)

        for epoch in range(1, self.epochs + 1):
            step = self.train_step(X, y, epoch)
            if step["diverged"]:
                break

        return self

    def predict(self, X: np.ndarray, price_stats: Dict[str, float]) -> np.ndarray:
        """Predicts in raw USD using the training price mean and standard deviation."""
        y_norm = self.predict_norm(X)
        return y_norm * price_stats["stdDev"] + price_stats["mean"]

    def evaluate(
        self,
        X_test: np.ndarray,
        y_test_norm: np.ndarray,
        y_test_raw: np.ndarray,
        price_stats: Dict[str, float]
    ) -> Dict[str, Any]:
        """Computes R², MSE, and RMSE on unseen test partition."""
        y_pred_norm = self.predict_norm(X_test)
        y_pred_raw = y_pred_norm * price_stats["stdDev"] + price_stats["mean"]

        # R² = 1 - (SS_res / SS_tot)
        ss_res = np.sum((y_test_raw - y_pred_raw) ** 2)
        ss_tot = np.sum((y_test_raw - np.mean(y_test_raw)) ** 2)
        r2 = float(1.0 - (ss_res / ss_tot)) if ss_tot > 0 else 0.0

        mse_raw = float(np.mean((y_test_raw - y_pred_raw) ** 2))
        rmse_raw = float(np.sqrt(mse_raw))
        mse_norm = float(np.mean((y_test_norm - y_pred_norm) ** 2))

        return {
            "r2": r2,
            "mse": mse_raw,
            "rmse": rmse_raw,
            "normalizedMSE": mse_norm,
            "weights": [float(w) for w in self.weights],
            "bias": float(self.bias),
            "history": self.history
        }
